feat: new composer

GeopJr opened the pull request:
fix: #553 
fix: #973
fix: #472
fix: #1453

GeopJr commented:
Judging by the errors, we are going to have to bump some dependencies, I'll make a last patch release when the current release is ready for the stable distros but that's it. I'm not going to hold back the new composer any longer and all the stuff used are **needed**.

bertob commented:
I tried the branch, mostly looks great to me! Thanks so much for working on this :raised_hands: 

A few minor issues I came across:

The alt text indicator can be hard to read depending on the image, perhaps we could make it more (or fully?) opaque? It also becomes more transparent on hover, which looks odd.

![image](https://github.com/user-attachments/assets/6399f3a8-b791-4c51-8fa8-afbea7441934)

![image](https://github.com/user-attachments/assets/15c8892e-1018-42e3-ba96-5e644b818436)


![image](https://github.com/user-attachments/assets/9d075dd1-caea-45b7-b539-f33f67d7d228)

The alt text below the image in the editor looks a bit lost layout-wise, since it's not anchored to the content. 

![Screenshot From 2025-06-22 12-02-23](https://github.com/user-attachments/assets/a7940e32-735c-4e21-aef0-310a82676b46)

At small sizes it's also a bit odd that the alt text scrolls on its own. I think it'd be better to scroll the entire view, including the image and the row below with the character count.

![image](https://github.com/user-attachments/assets/00288c71-8902-4584-bbec-ed3646c7411f)

It's not ideal that the single choice / show results labels are ellipsized. I realize now that this was a flaw in the mockups, since I never actually tried drawing this for phone sizes. Maybe we could just move the third item to a second row and make it full-width though?

![Screenshot From 2025-06-22 12-22-19](https://github.com/user-attachments/assets/598add92-0a5e-4f8f-b46e-3f8f078f411f)

When opening the scheduling dialog it has a different width than the main post dialog:

[Screencast From 2025-06-22 12-47-39.webm](https://github.com/user-attachments/assets/89c754eb-a6a4-4a10-9e7e-ee14b8715bbc)

GeopJr commented:
Thanks for the review :pray: 

> The alt text indicator can be hard to read depending on the image, perhaps we could make it more (or fully?) opaque? It also becomes more transparent on hover, which looks odd.


| | |
|---|---|
| ![Screenshot From 2025-06-22 14-01-36](https://github.com/user-attachments/assets/46f7de73-81a1-4d4e-a5bb-1841448f1ec0) | ![Screenshot From 2025-06-22 14-01-26](https://github.com/user-attachments/assets/37b699ea-d0e1-4d64-ba26-8a0ddbcd84ca) |
| ![Screenshot From 2025-06-22 14-38-15](https://github.com/user-attachments/assets/ae1a461b-466a-45f0-a9dc-71c63bd6b072) | ![Screenshot From 2025-06-22 14-37-57](https://github.com/user-attachments/assets/8b5a7eda-2be3-41e1-9825-351af6615984) |

I think the second row is better IMO, more opaque + blended with white/black while keeping the text color. (FWIW, in high contrast the indicator becomes completely opaque etc.)

> It also becomes more transparent on hover, which looks odd.

My reasoning was that since it's not a button, having it become more transparent would better show that it's just an indicator and by hovering the user can see the attachment below it better. Not strong opinion on it, I'll remove it!

> The alt text below the image in the editor looks a bit lost layout-wise, since it's not anchored to the content.

Can you expand on what you mean by 'anchored to the content'? / How it can be fixed?

> At small sizes it's also a bit odd that the alt text scrolls on its own. I think it'd be better to scroll the entire view, including the image and the row below with the character count.

I agree on the character counter but I'm not so sure about scrolling with the attachment. I feel that if you are writing alt text for an image, you'd want to keep looking at it instead of scrolling up and down to see and describe it. Mastodon mobile seems to do that too, the image can be scrolled but the alt text box stays in place at the bottom.
(The character counter could also be moved to the headerbar or inside the image as an overlay)

> It's not ideal that the single choice / show results labels are ellipsized. I realize now that this was a flaw in the mockups, since I never actually tried drawing this for phone sizes. Maybe we could just move the third item to a second row and make it full-width though?

||||
|---|---|---|
| ![Screenshot From 2025-06-22 14-23-56](https://github.com/user-attachments/assets/29daa7ef-dc7e-4c8e-a60c-94a3e441ffb1) | ![Screenshot From 2025-06-22 14-23-32](https://github.com/user-attachments/assets/b0edb96b-4531-4300-af71-f932284798d7) | ![Screenshot From 2025-06-22 14-22-42](https://github.com/user-attachments/assets/a78d4c3d-5f72-41ff-8983-c17f4ba60e77) |

I think it looks a bit awkward on the non-bottom-sheet versions. Maybe keep it as 3 columns there? It would ellipsize though... Alternatively I can increase the default window width.

GeopJr commented:
- Attachment editor

Added a light shade background + paddings and added a clamp for the alt text. Not sure if it's a good idea however and might be better to do only horizontal margins. Counter moved to the headerbar. FWIW, videos and audio show a video player instead of a focus picker.

|||
|---|---|
| ![Screenshot From 2025-06-24 04-15-06](https://github.com/user-attachments/assets/d8a7ac7b-7f31-4e03-8733-a74edf07b68e) | ![Screenshot From 2025-06-24 04-15-00](https://github.com/user-attachments/assets/a7444c4b-5219-4db0-97c5-1b910d2808cb) |

![Screenshot From 2025-06-24 04-15-48](https://github.com/user-attachments/assets/ecb73951-962e-4681-afe9-aefd51ed5ba0)


- Polls

|||
|---|---|
| ![Screenshot From 2025-06-24 04-25-22](https://github.com/user-attachments/assets/c8a6aa10-f73e-481f-b20c-77a1dad31c80) | ![Screenshot From 2025-06-24 04-25-02](https://github.com/user-attachments/assets/40e489fc-568f-4a58-bae8-a22feb67cc20) |
| ![Screenshot From 2025-06-24 04-24-16](https://github.com/user-attachments/assets/fba8c51c-bae9-4958-b28f-9600a5e0a749) | ![Screenshot From 2025-06-24 04-29-43](https://github.com/user-attachments/assets/9f2f6651-131d-4869-8b32-5b9260633a3d) |
| ![Screenshot From 2025-06-24 04-49-46](https://github.com/user-attachments/assets/336f35ff-66ea-4629-a17a-fe860db5f50b) | ![Screenshot From 2025-06-24 04-49-19](https://github.com/user-attachments/assets/b70ae1bd-0375-431e-8ea8-a699e6aa3cfe) |

The first three are no-arrow with different alignments, the fourth one is no-arrow and short label and the other two are icon-only show results with different alignments.

I kind-of see only the last two being the least awkward... I don't think 3 long-ish labels will fit nicely in that space and it might be even more jarring with other languages that tend to have longer words :/

---

Pushed the better colored indicators (and implemented the validity checks so the 'success' variant can be tested now). I did notice **two things I missed**:

1. The old composer has a button for marking media as sensitive which the mockups don't. Should I add it to the bottom bar? header bar? below the attachments?
2. the mockups have a unique symbolic for the polls which is not present in library, if you have it at hand, please send it!

bertob commented:
### Image editor

> Added a light shade background + paddings and added a clamp for the alt text. Not sure if it's a good idea however and might be better to do only horizontal margins. Counter moved to the headerbar. FWIW, videos and audio show a video player instead of a focus picker.

It looks a bit odd that the text is not aligned with the image, maybe from that PoV it'd be better to just do full-width? The horizontal padding is also a bit too wide now, I'd go with a similar horizontal and vertical padding on the textarea (e.g. 12px).

I'd also increase the vertical padding around the image to 12, it looks a bit cramped currently.

As for the counter, on the top left it looks like it belongs to the back button, which is not ideal. Perhaps we could drop the view title and move it to the right? Could also try dropping the bold and maybe making it dimmed, to make it clearer that it's not a button. I'd also add a tooltip like `Remaining Characters`.

![image](https://github.com/user-attachments/assets/462d9ef0-6cb6-4131-852e-21ff75e6339c)

More generally, it's a bit of an odd pattern to have a sub-view with a back button that deletes all changes unless you hit "save". Is there a reason why we can't just insta-apply the changes without the need for a save button?

### Polls

I'd try going with this for the narrow case, or maybe this but with the date thing using half the width and left-aligned:

![image](https://github.com/user-attachments/assets/ac1d381f-9ec8-433f-afd9-ff5c0610de07)

and this (with the arrow) for the wide case - maybe with an even narrower duration picker, it's fine if that one is ellipsized a bit in some cases since it'd just be the last few characters of "minutes":

![image](https://github.com/user-attachments/assets/bb072e4f-488a-4d23-adf0-66a31a0c642c)

---

> The old composer has a button for marking media as sensitive which the mockups don't. Should I add it to the bottom bar? header bar? below the attachments?

In the old composer it looks like it's per-post rather than per-image, is that how the setting works? Otherwise maybe it could be an overlay with an eye icon in the top left of the thumbnails?

> the mockups have a unique symbolic for the polls which is not present in library

![polls-symbolic](https://github.com/user-attachments/assets/4137d69e-ff3c-470f-84ed-5a28adaddcfe)

bertob commented:
### Image Editor
The latest version looks good to me, though I would add back a title (maybe just "Edit Media") since we have enough space now.

### Saving to server
More of a general point and I'm not sure how actionable it is, but I'm really not sure about sending anything to the server until you actually submit the post (I just realized now that images are actually immediately uploaded when you add them). From a user point of view this seems counterintuitive in a native app (rather than a website, where the assumption is that anything you do goes to the server).

No idea what our options are here, but maybe worth considering saving drafts locally only until you actually send/explicitly save as draft on the server?

### Polls
That video doesn't work for me in multiple (Flatpak, so not a codec thing I think?) browsers and downloading doesn't work either :(

### CW for media
Can we just add another icon button (maybe with the strikethrough eye) next to the other CW button then?

GeopJr commented:
# Saving to server

It could be possible for the alt text + focus but the media is a sticky one. When we send the final post to the server we attach the uploaded media ids, which means we first have to upload them separately. The issue is that uploading can fail for a number of reasons outside of our control (we already check for size and what's allowed etc). For example Mastodon often drops (dropped?) videos that had weird framerates. So the user would have to deal with these when pressing 'Post'. Other than that, there's some strict rate limits around media, so uploading them all at once can reach them fast (often what happens on the tuba@floss.social announcements and the threads stay unfinished until the limits pass).

I'll do a bit more research on this but I'm not too positive it will work reliably.

# Polls

![ScreencastFrom2025-06-2705-48-59-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/530ef605-6091-4944-bcce-134f4320add5)

# CW for media

It should only be visible when there are attachments, right?

bertob commented:
I tried the grid options, none of them are great but the left aligned one is the least bad one for the narrow layout. I think we could go with that for mobile, but then switch to the three column layout (with a narrow third column) as soon as possible. Maybe we could also try to switch to shorter strings in the three column case, e.g. using "min" instead of "Minutes" and "h" instead of "Hours", to minimize the width of the third column.

![image](https://github.com/user-attachments/assets/5f51a642-4f24-4de3-ac8b-06740947da60)

> It should only be visible when there are attachments, right?

Yup, that's what I was thinking.

bertob commented:
The poll thing looks good to me layout wise, but I'd go with more standard abbreviations, i.e. lowercase letter for hours and days, and not abbreviate minutes to "M". So e.g. `5min`, `6h`, `3d`.

The sensitive media setting looks good, but it'd be nicer if it actually previewed that the images will be blurred. Could we already blur the thumbnails in the composer when the setting is on to make it more obvious what it does?

bertob commented:
LGTM :)
